File: Segundo_Curso/Segundo_Cuatri/PDOO/ruby/Irrgarten/Monster.rb
#encoding: UTF-8

require_relative 'Dice'
require_relative 'LabyrinthCharacter'


module Irrgarten

    class Monster < LabyrinthCharacter

        #Vida inicial de un monstruo
        @@INITIAL_HEALTH = 5

        def initialize(name, intelligence, strength)
            super(name,intelligence,strength,@@INITIAL_HEALTH)
        end

        #TODO*: Revisar si todos a gusto
        #*Eliminados: dead, setPos, gotWounded

        #Calcula la intensidad de ataque del monstruo
        def attack
            atk=0
            atk=Dice.intensity(@strength) unless dead
            atk
        end

        #Devuelve el estado del monstruo en forma de cadena
        def to_s
            "M"+super
        end

        #Calcula como se defiende el monstruo
        def defend(receivedAttack)
		    isDead = dead
          	if !isDead
                defensiveEnergy = Dice.intensity(@intelligence)
                if(defensiveEnergy < receivedAttack)
                    gotWounded
                    isDead = dead
                end
		    end
		    isDead
        end
    end#class
end#module

File: Segundo_Curso/Segundo_Cuatri/PDOO/java/src/irrgarten/Monster.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package irrgarten;

/**
 *
 * @author el_dramas
 */
public class Monster extends LabyrinthCharacter {
    
    static private final int INITIAL_HEALTH = 5; 
    
    /** 
     * Crea un monstruo
     * @param name Su nombre
     * @param intelligence Su defensa / inteligencia
     * @param strength su fuerza
     */
    public Monster (String name, float intelligence, float strength){
        super (name,intelligence,strength,INITIAL_HEALTH);
    }
    
    /**
     * Comprueba si un monstruo esta muerto
     * @return True si el monstruo esta muerto
     */
    //@Override
    /*public boolean dead(){
        return super.dead();
    } 
    
    /**
     * Maneja un ataque
     * @return el valor del ataque
     */
    @Override
    public float attack(){
        if (this.dead())
            return 0;
        else return Dice.intensity(super.getStrength());
    }
    
    /**
     * Maneja la defensa
     * @param receivedAttack la fuerza del ataque recibido
     * @return True si muere con el ataque
     */
    @Override
    public boolean defend(float receivedAttack){
        Boolean isDead = dead();
        if(!isDead){
            float defensiveEnergy = Dice.intensity(super.getIntelligence());
            if(defensiveEnergy < receivedAttack){
                gotWounded();
                isDead = dead();
            }
        }
        return isDead;
    }
    
    /**
     * Guarda la posicion del monstruo
     * @param row La fila
     * @param col La columna
     */
    @Override
    public void setPos(int row, int col){
        super.setPos(row, col);
    }
    
    @Override
    public String toString(){
        return "M"+super.toString(); 
    }
    
}